containers/ruby-consul-terraform/render_configs.rb

```diff
@@ -0,0 +1,125 @@
+  manifest = "base-configs/#{$app_name}/manifest.rb"
+  overwrite_prompt = true
+  $instance_name = instance_name
+  render_dir = "/data/configs/#{instance_name}"
+  FileUtils.mkdir_p(render_dir) unless File.exists?(render_dir)
+  $base_dir = render_dir
+  $output_dir = "#{render_dir}/configs"
+  Dir.chdir(render_dir) do
+    configure(manifest, overwrite_prompt)
+  end
+  Open3.popen3("gsutil rsync -r -d #{render_dir} gs://#{config_bucket}/#{instance_name}/") { |stdin, stdout, stderr, wait_thread|
+    puts stdout.read
+    puts stderr.read
+  }
+end
+
+if __FILE__ == $0
+  $base_dir = Dir.pwd
+  copy_file_from_github "configure.rb"
+  instances = JSON.parse(ENV.fetch("INSTANCES", "")).map {|self_link| self_link.split('/')[-1]}
+  ENV["OUTPUT_DIR"] = "/dev/null" # This must be set when configure.rb is loaded but is not used
+  require "/data/configure.rb"
+
+  # Ugly hack -- the `configure.rb` script from the repo uses a function
+  # called `render_ctmpl` that calls a docker container to run consul-template
+  # to render files. Because we're running this script in a docker container,
+  # we can't use docker that way. So instead, bake consul-template into this
+  # container and override the `render_ctmpl` function with one that uses
+  # consul-template locally.
+  def render_ctmpl(file_name, output_file_name, tmp_dir=nil)
+    if tmp_dir.nil?
+      tmp_dir = (Dir.pwd.split("/") - $base_dir.split("/")) * "/"
+    end
+    cmd_env = {
+      "VAULT_TOKEN" => ENV.fetch("VAULT_TOKEN", ""),
+      "INSTANCE_TYPE" => $instance_type,
+      "RUN_CONTEXT" => $run_context,
+      "APP_NAME" => $app_name,
+      "ENVIRONMENT" => $env,
+      "INSTANCE_NAME" => $instance_name,
+      "HOST_TAG" => $host_tag,
+      "IMAGE" => $image,
+      "TARGET_DOCKER_VERSION" => $target_docker_version,
+      "VAULT_ADDR" => $vault_addr,
+      "GOOGLE_PROJ" => $google_proj,
+      "GOOGLE_APPS_DOMAIN" => $apps_domain,
+      "GOOGLE_APPS_ORGANIZATION_ID" => $apps_organization_id,
+      "GOOGLE_APPS_SUBDOMAIN" => $apps_subdomain,
+      "GCS_NAME_PREFIX" => $gcs_name_prefix,
+      "DNS_DOMAIN" => $dns_domain,
+      "LDAP_BASE_DOMAIN" => $ldap_base_domain,
+      "BUCKET_TAG" => $bucket_tag,
+      "SERVICE_VERSION" => ENV.fetch("SERVICE_VERSION", ""),
+      "DIR" => $fiab_dir
+    }
+    Open3.popen3(
+      cmd_env, 
+      "consul-template", 
+      "-config=#{$vault_config_path}", 
+      "-template=/data/configs/#{$instance_name}/#{tmp_dir}/#{file_name}:/data/configs/#{$instance_name}/#{tmp_dir}/#{output_file_name}",
+      "-once"
+    ) { |stdin, stdout, stderr, wait_thread|
+    if wait_thread.value.success?
+      puts "#{file_name} > #{output_file_name}"
+      File.delete(file_name)
+    else
+      puts stderr.read
+      $failure_rendering = true
+      $failed_to_render_file_names.push(file_name)
+    end
+    }
+  end
+  activate_sa("/data/provider_sa.json")
+  instances.map {|i| render_instance_configs(i, ENV.fetch("CONFIG_BUCKET")) }
+end
```

Comment:
this duplicates a lot of what's in dsde-toolbox. Is there a way to reuse, instead of copy-pasting?

Reply:
Interesting point. I didn't realize it was in dsde-toolbox; I was looking at the [version in firecloud-develop](https://github.com/broadinstitute/firecloud-develop/blob/dev/configure.rb) which is used in the [current render process](https://github.com/broadinstitute/firecloud-develop/blob/dev/jenkins/jenkins-instance-deploy.sh#L161) on instances. Those two versions of `configure.rb` have significant differences; the one in dsde-toolbox appears not to be as recently maintained.

I thought I needed to copy/paste the `copy_file_from_github` function (because I "needed to get configure.rb"). I also thought I needed the `set_github_token` function for the same reason, though it seems like the existing `read_secret_from_path` function would work.

So it _does_ seem like I could probably salvage stuff from dsde-toolbox, but it would be pretty hacky. In order to maintain compatibility with the current render process, I want to use the `configure.rb` from firecloud develop for the actual render, because that's what the manifest files in firecloud-develop are written to. But I could require the `configure.rb` that's already in the container, use the `copy_file_from_github` and `set_github_token` from there to copy down the `configure.rb` from firecloud-develop, _then_ require the `configure.rb` from firecloud develop, sort out any differences in the environment that arise from the two configure scripts disagreeing about environment variables (I noticed `ENV` vs `ENVIRONMENT`; there may be others) and then continue with the redefinition of `render_ctmpl` and doing the configure. But that seems like a lot of indirection for the amount of duplicated code we'd save.

Reply:
ah - then this is my bad too, I didn't realize which `configure.rb` was actually in use. I'm not at all attached to the firecloud-develop version or the dsde-toolbox version. My point here is really "reuse, don't copy-paste"  as  much as  possible.